subt/tools/ignstate.py

```diff
@@ -0,0 +1,170 @@
+""""
+  Extract robot positions from Ignition SQLite log file
+"""
+import cv2
+import numpy as np
+import re
+import sqlite3
+import xml.etree.ElementTree as ET
+
+from datetime import timedelta
+
+from ignition.msgs.pose_v_pb2 import Pose_V
+
+COLORS = [
+    (255, 255, 255),
+    (0, 255, 0),
+    (0, 0, 255),
+    (255, 0, 0),
+    (0, 255, 255),
+    (255, 0, 255),
+    (255, 255, 0),
+]
+SCALE = 10  # 1 pixel is 1dm
+BORDER_PX = 10  # extra border
+
+MARKERS = {
+    'backpack': (cv2.MARKER_SQUARE, 3),
+    'rescue_randy': (cv2.MARKER_DIAMOND, 2),
+    'gas': (cv2.MARKER_TRIANGLE_UP, 1),
+    'vent': (cv2.MARKER_STAR, 4),
+    'phone': (cv2.MARKER_TILTED_CROSS, 5),
+    'artifact_origin': (cv2.MARKER_CROSS, 6),
+}
+
+
+def read_poses(filename, seconds=3700):
+    ret = []
+
+    con = sqlite3.connect(filename)
+    cursor = con.cursor()
+    cursor.execute(r"SELECT id FROM topics where name LIKE '%/dynamic_pose/info';")
+    result = cursor.fetchone()
+    dynamic_topic_id = result[0]
+
+    poses = Pose_V()
+    try:
+        # cannot use WHERE filtering since the state.log is always corrupted
+        cursor.execute("SELECT message, topic_id FROM messages")
+        for m, topic_id in cursor:
+            if topic_id != dynamic_topic_id:
+                continue
+            poses.ParseFromString(m)
+            timestamp = timedelta(seconds=poses.header.stamp.sec, microseconds=poses.header.stamp.nsec / 1000)
+            if timestamp > timedelta(seconds=seconds):
+                return ret
+            current = dict()
+            for pose in poses.pose:
+                if "_" in pose.name:
+                    continue
+                current[pose.name] = pose.position
+            if len(current) > 0:
+                ret.append((timestamp, current))
+    except sqlite3.DatabaseError as e:
+        print(f"{type(e).__name__}: {e}")
+
+    return ret
+
+
+def read_artifacts(filename):
+    ret = []
+    con = sqlite3.connect(filename)
+    cursor = con.cursor()
+    cursor.execute(f"SELECT id FROM topics WHERE name == '/logs/sdf'")
+    result = cursor.fetchone()
+    sdf_id = result[0]
+    cursor.execute(r"SELECT message, topic_id FROM messages")
+    for message, topic_id in cursor:
+        if topic_id == sdf_id:
+            world = message
+            break
+    root = ET.fromstring(world[4:])
+    type_re = re.compile('^(backpack|rescue_randy|gas|vent|phone|artifact_origin)')
```

Comment:
do we want to support also Tunnel Circuit artifacts? (Drill, Fire Extinguisher) ... or "later" when Cave artifacts will be available?

Reply:
Yes, I think we do. But I'd like to keep this PR to be urban-only since adding other things would require fairly substantial retesting on data I currently don't have readily available.